BeanArchiveProcessor#build threw an exception: Local name must not be null

### Describe the bug

I am wanting to migrate a large existing codebase from Tomcat to Quarkus. When I do `./gradlew :quarkusAppPartsBuild` I get this error:

```
2024-10-11 12:53:42,970 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.BeanArchiveProcessor#build threw an exception: java.lang.IllegalArgumentException: Local name must not be null
        at org.jboss.jandex.DotName.<init>(DotName.java:137)
        at org.jboss.jandex.DotName.createSimple(DotName.java:77)
        at io.quarkus.deployment.index.IndexingUtil.indexClass(IndexingUtil.java:151)
        at io.quarkus.arc.deployment.BeanArchiveProcessor.build(BeanArchiveProcessor.java:94)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:252)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:58)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:113)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:435)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:56)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.BeanArchiveProcessor#build threw an exception: java.lang.IllegalArgumentException: Local name must not be null
...
```

I can see from the debugger that the null name comes from the list of additional bean classes, but I don’t know where that list comes from.

How would I go about diagnosing this error?

### Expected behavior

The build succeeds.

### Actual behavior

The build fails.

### How to Reproduce?

I doubt that I could reproduce this outside of my existing codebase, not without knowing what the issue is.

### Output of `uname -a` or `ver`

Darwin Rods-MacBook-Pro.local 24.0.0 Darwin Kernel Version 24.0.0: Tue Sep 24 23:35:10 PDT 2024; root:xnu-11215.1.12~1/RELEASE_ARM64_T6031 arm64

### Output of `java -version`

openjdk version "21.0.4" 2024-07-16 OpenJDK Runtime Environment Homebrew (build 21.0.4) OpenJDK 64-Bit Server VM Homebrew (build 21.0.4, mixed mode, sharing)

### Quarkus version or git rev

3.15.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.9

### Additional information

_No response_

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
@@ -430,10 +430,14 @@ public class UndertowBuildStep {
         //add servlets
         if (webMetaData.getServlets() != null) {
             for (ServletMetaData servlet : webMetaData.getServlets()) {
+                String servletClass = servlet.getServletClass();
+                if (servletClass == null) {
+                    continue;
+                }
                 reflectiveClasses.accept(
-                        ReflectiveClassBuildItem.builder(servlet.getServletClass()).build());
+                        ReflectiveClassBuildItem.builder(servletClass).build());
                 RuntimeValue<ServletInfo> sref = recorder.registerServlet(deployment, servlet.getServletName(),
-                        context.classProxy(servlet.getServletClass()),
+                        context.classProxy(servletClass),
                         servlet.isAsyncSupported(),
                         servlet.getLoadOnStartupInt(),
                         bc.getValue(),
@@ -447,7 +451,7 @@ public class UndertowBuildStep {
                 if (webMetaData.getAnnotations() != null) {
                     for (AnnotationMetaData amd : webMetaData.getAnnotations()) {
                         final ServletSecurityMetaData ssmd = amd.getServletSecurity();
-                        if (ssmd != null && amd.getClassName().equals(servlet.getServletClass())) {
+                        if (ssmd != null && amd.getClassName().equals(servletClass)) {
                             // Process the @ServletSecurity into metadata
                             ServletSecurityInfo securityInfo = new ServletSecurityInfo();
                             securityInfo.setEmptyRoleSemantic(
@@ -471,7 +475,7 @@ public class UndertowBuildStep {
                         }
 
                         final MultipartConfigMetaData mcmd = amd.getMultipartConfig();
-                        if (mcmd != null && amd.getClassName().equals(servlet.getServletClass())) {
+                        if (mcmd != null && amd.getClassName().equals(servletClass)) {
                             servlet.setMultipartConfig(mcmd);
                         }
                     }

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/WebXmlParsingBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/WebXmlParsingBuildStep.java
@@ -79,7 +79,10 @@ public class WebXmlParsingBuildStep {
                     }
                     if (metadata.getServlets() != null) {
                         for (ServletMetaData i : metadata.getServlets()) {
-                            additionalBeans.add(i.getServletClass());
+                            String servletClass = i.getServletClass();
+                            if (servletClass != null) {
+                                additionalBeans.add(servletClass);
+                            }
                         }
                     }
                     if (metadata.getFilters() != null) {

